cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,543 @@
+   * @return True if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * false otherwise.
+   */
+  @Override
+  protected boolean selfSatisfied(ClusterModel clusterModel, BalancingAction action) {
+    if (!topicInterestedByThisGoal(action.topic()) || !topicsWithMinLeadersPerBrokerNames(clusterModel).contains(action.topic())) {
+      // Always self-satisfied on topics that are not of the interest of this goal
+      return true;
+    }
+    switch (action.balancingAction()) {
+      case LEADERSHIP_MOVEMENT:
+      case INTER_BROKER_REPLICA_MOVEMENT:
+        Broker sourceBroker = clusterModel.broker(action.sourceBrokerId());
+        Replica replicaToBeMoved = sourceBroker.replica(action.topicPartition());
+        if (!replicaToBeMoved.isLeader()) {
+          return true; // Move a follower replica of a topic of this goal's interest is always self-satisfied since it does not matter
+        }
+        return isSelfSatisfiedToMoveOneLeaderReplica(sourceBroker,
+                                                     clusterModel.broker(action.destinationBrokerId()),
+                                                     action.topic());
+      case INTER_BROKER_REPLICA_SWAP:
+        Replica sourceReplica = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+        Replica destinationReplica = clusterModel.broker(action.destinationBrokerId()).replica(action.topicPartition());
+        if (sourceReplica.isLeader() == destinationReplica.isLeader()) {
+          // Swapping either 2 leader replicas or 2 follower replicas has no effect on this goal
+          return true;
+        }
+        // At this point, one replica must be leader and the other must be follower
+        if (sourceReplica.isLeader()) {
+          return isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.sourceBrokerId()),
+                                                       clusterModel.broker(action.destinationBrokerId()),
+                                                       action.topic());
+        }
+        // The replica on the destination broker is a leader replica
+        return isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.destinationBrokerId()),
+                                                     clusterModel.broker(action.sourceBrokerId()),
+                                                     action.topic());
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private boolean topicInterestedByThisGoal(String topicName) {
+    if (_balancingConstraint.topicsWithMinLeadersPerBrokerPattern().pattern().isEmpty()) {
+      return false; // No topic is of this goal's interest
+    }
+    return _balancingConstraint.topicsWithMinLeadersPerBrokerPattern().matcher(topicName).matches();
+  }
```

Comment:
From our earlier discussion, I remember that doing a pattern match is an expensive operation. Can be more efficient by doing the following:
1. In `initGoalState` method of this goal, calculate the set of all topics relevant to this goal (see `KafkaCruiseControl#excludedTopics(ClusterModel, Pattern)` regarding how this can ben done),
2. then keep the result in a local hashset.
3. In places, where the check is relevant (e.g. in `actionAcceptance`), just use set,contains().

Reply:
Good idea